agent/consul/gateways/controller_gateways.go

```diff
@@ -840,151 +828,132 @@ func newGatewayMeta(gateway *structs.APIGatewayConfigEntry, bound structs.Config
+	return structs.NewGatewayCondition(
+		structs.GatewayConditionAccepted,
+		structs.ConditionStatusTrue,
+		structs.GatewayReasonAccepted,
+		"gateway is valid",
+		structs.ResourceReference{},
+	)
 }
 
 // invalidCertificate returns a condition used when a gateway references a
 // certificate that does not exist. It takes a ref used to scope the condition
 // to a given APIGateway listener.
-func (g *gatewayConditionGenerator) invalidCertificate(ref structs.ResourceReference, err error) structs.Condition {
-	return structs.Condition{
-		Type:               "Accepted",
-		Status:             "False",
-		Reason:             "InvalidCertificate",
-		Message:            err.Error(),
-		Resource:           pointerTo(ref),
-		LastTransitionTime: g.now,
-	}
+func invalidCertificate(ref structs.ResourceReference, err error) structs.Condition {
+	return structs.NewGatewayCondition(
+		structs.GatewayConditionResolvedRefs,
+		structs.ConditionStatusFalse,
+		structs.GatewayListenerReasonInvalidCertificateRef,
+		err.Error(),
+		ref,
+	)
 }
 
 // invalidCertificates is used to set the overall condition of the APIGateway
 // to invalid due to missing certificates that it references.
-func (g *gatewayConditionGenerator) invalidCertificates() structs.Condition {
-	return structs.Condition{
-		Type:               "Accepted",
-		Status:             "False",
-		Reason:             "InvalidCertificates",
-		Message:            "gateway references invalid certificates",
-		LastTransitionTime: g.now,
-	}
+func invalidCertificates() structs.Condition {
+	return structs.NewGatewayCondition(
+		structs.GatewayConditionResolvedRefs,
+		structs.ConditionStatusFalse,
+		structs.GatewayListenerReasonInvalidCertificateRef,
+		"gateway references invalid certificates",
+		structs.ResourceReference{},
+	)
 }
 
-// gatewayAccepted marks the APIGateway as valid.
-func (g *gatewayConditionGenerator) gatewayAccepted() structs.Condition {
-	return structs.Condition{
-		Type:               "Accepted",
-		Status:             "True",
-		Reason:             "Accepted",
-		Message:            "gateway is valid",
-		LastTransitionTime: g.now,
-	}
+// gatewayListenerNoConflicts marks an APIGateway listener as having no conflicts within its
+// bound routes
+func gatewayListenerNoConflicts(ref structs.ResourceReference) structs.Condition {
+	return structs.NewGatewayCondition(
+		structs.GatewayConditionListenersConfigured,
+		structs.ConditionStatusTrue,
+		structs.GatewayReasonListenersConfigured,
+		"listener has no route conflicts",
+		ref,
+	)
+}
+
+// gatewayListenerConflicts marks an APIGateway listener as having bound routes that conflict with each other
+// and make the listener, therefore invalid
+func gatewayListenerConflicts(ref structs.ResourceReference) structs.Condition {
+	return structs.NewGatewayCondition(
+		structs.GatewayConditionListenersConfigured,
+		structs.ConditionStatusFalse,
+		structs.GatewayListenerReasonProtocolConflict,
```

Comment:
Also, this appears to change things to say `ProtocolConflict` which actually isn't accurate here -- there's no "protocol conflict" that we can ever have as far as a listener is concerned. In particular this helper is used due to the fact that we don't support TCP load-balancing, so if you bind multiple routes to a TCP-based listener we say "this thing is invalid". So I think we'll want to stick with the custom reason here.